source/simulator/helper_functions.cc

```diff
@@ -909,8 +909,134 @@ namespace aspect
 
     return (material_model->get_model_dependence().viscosity != MaterialModel::NonlinearDependence::none);
   }
-}
 
+
+
+  template <int dim>
+  void Simulator<dim>::apply_limiter_to_dg_solutions (LinearAlgebra::BlockVector &vector)
+  {
+    // create a quadrature formula based on the compositional element alone.
+    // be defensive about determining that a compositional field actually exists
+    if ( parameters.n_compositional_fields == 0)
+      AssertThrow (introspection.base_elements.compositional_fields
+                   != numbers::invalid_unsigned_int,
+                   ExcMessage("This function cannot be used without compositional fields."));
+
+    const QGauss<1> quadrature_formula_1(parameters.composition_degree+1);
+    const QGaussLobatto<1> quadrature_formula_2(parameters.composition_degree+1);
+    /**
+     * TODO: currently it only works for 2 dimentional problem. Will add the 3D special selected quadrature points and weight later.
+     * Construct the quadrature points and weights:
+     * x direction uses Gauss points and y direction uses Gauss Lobatto points
+     */
+    const QAnisotropic<dim> quadrature_formula_12(quadrature_formula_1,quadrature_formula_2);
+    //Cddonstruct the quadrature points and weights:
+    //x direction uses Gauss Lobatto points and y direction uses Gauss points
+    const QAnisotropic<dim> quadrature_formula_21(quadrature_formula_2,quadrature_formula_1);
+    const unsigned int n_q_points_12 = quadrature_formula_12.size();
+    const unsigned int n_q_points_21 = quadrature_formula_21.size();
+    FEValues<dim> fe_values_12 (mapping,
+                                finite_element,
+                                quadrature_formula_12,
+                                update_values   |
+                                update_quadrature_points |
+                                update_JxW_values);
+    std::vector<double> compositional_values_12(n_q_points_12);
+    FEValues<dim> fe_values_21 (mapping,
+                                finite_element,
+                                quadrature_formula_21,
+                                update_values   |
+                                update_quadrature_points |
+                                update_JxW_values);
+    std::vector<double> compositional_values_21(n_q_points_21);
+
+    std::vector<double> local_compositional_average_integrals (parameters.n_compositional_fields);
+
+    const double composition_exact_max = 1.0;//TODO: need to add into the parameters list, let the user to define.
```

Comment:
I would suggest to introduce the new parameters in the "Discretization/Stabilization parameters" subsection. I could also imagine moving the "Use limiter for discontinuous solution" parameter into that section, since it essentially is not a property of the discretization, but a postprocess to stabilize it (like the artificial viscosity or viscosity smoothing in the continuous case, and the discontinuous penalty parameter is also already in that subsection). I would leave the decision about the "Use limiter for discontinuous solution" parameter up to you, but the bounds should definitely go in the subsection.